cpp/src/arrow/util/align_util_test.cc

```diff
@@ -278,4 +279,225 @@ TEST(EnsureAlignment, Table) {
+std::shared_ptr<DataType> sample_type<Decimal128Type>() {
+  return decimal128(32, 6);
+}
+
+template <>
+std::shared_ptr<DataType> sample_type<Decimal256Type>() {
+  return decimal256(60, 10);
+}
+
+template <>
+std::shared_ptr<DataType> sample_type<LargeListType>() {
+  return large_list(int8());
+}
+
+template <>
+std::shared_ptr<DataType> sample_type<DenseUnionType>() {
+  return dense_union({field("x", int8()), field("y", uint8())});
+}
+
+template <>
+std::shared_ptr<DataType> sample_type<MapType>() {
+  return map(utf8(), field("item", utf8()));
+}
+
+template <>
+std::shared_ptr<DataType> sample_type<DurationType>() {
+  return duration(TimeUnit::NANO);
+}
+
+template <>
+std::shared_ptr<DataType> sample_type<TimestampType>() {
+  return timestamp(TimeUnit::NANO);
+}
+
+template <>
+std::shared_ptr<DataType> sample_type<Time32Type>() {
+  return time32(TimeUnit::SECOND);
+}
+
+template <>
+std::shared_ptr<DataType> sample_type<Time64Type>() {
+  return time64(TimeUnit::NANO);
+}
+
+template <>
+std::shared_ptr<DataType> sample_type<SparseUnionType>() {
+  return sparse_union({field("x", uint8()), field("y", int8())}, {1, 2});
+}
+
+template <typename ArrowType>
+std::shared_ptr<ArrayData> SampleArray() {
+  random::RandomArrayGenerator gen(42);
+  return gen.ArrayOf(sample_type<ArrowType>(), 100)->data();
+}
+
+template <>
+std::shared_ptr<ArrayData> SampleArray<SparseUnionType>() {
+  auto ty = sparse_union({field("ints", int64()), field("strs", utf8())}, {2, 7});
+  auto ints = ArrayFromJSON(int64(), "[0, 1, 2, 3]");
+  auto strs = ArrayFromJSON(utf8(), R"(["a", null, "c", "d"])");
+  auto ids = ArrayFromJSON(int8(), "[2, 7, 2, 7]")->data()->buffers[1];
+  const int length = 4;
+  SparseUnionArray arr(ty, length, {ints, strs}, ids);
+  return arr.data();
+}
+
+class MallocAlignment : public ::testing::Test {
+ public:
+  void CheckModified(const ArrayData& src, const ArrayData& dst) {
+    ASSERT_EQ(src.buffers.size(), dst.buffers.size());
+    for (std::size_t i = 0; i < src.buffers.size(); i++) {
+      if (!src.buffers[i] || !dst.buffers[i]) {
+        continue;
+      }
+      if (src.buffers[i]->address() != dst.buffers[i]->address()) {
+        return;
+      }
+    }
+    FAIL() << "Expected at least one buffer to have been modified by EnsureAlignment";
+  }
+
+  void CheckUnmodified(const ArrayData& src, const ArrayData& dst) {
+    ASSERT_EQ(src.buffers.size(), dst.buffers.size());
+    for (std::size_t i = 0; i < src.buffers.size(); i++) {
+      if (!src.buffers[i] || !dst.buffers[i]) {
+        continue;
+      }
+      ASSERT_EQ(src.buffers[i]->address(), dst.buffers[i]->address());
+    }
+  }
+};
+
+template <typename T>
+class MallocAlignmentRequired : public MallocAlignment {};
+template <typename T>
+class MallocAlignmentNotRequired : public MallocAlignment {};
+
+TYPED_TEST_SUITE(MallocAlignmentRequired, TypesRequiringSomeKindOfAlignment);
+TYPED_TEST_SUITE(MallocAlignmentNotRequired, TypesNotRequiringAlignment);
+
```

Comment:
Why not also a test that aligned values do not need any realignment, or did I miss an existing test for this?

Reply:
Something like:
```c++
TYPED_TEST(MallocAlignmentRequired, AlreadyAligned) {
  std::shared_ptr<ArrayData> data = SampleArray<TypeParam>();
  ASSERT_OK_AND_ASSIGN(
      std::shared_ptr<ArrayData> aligned,
      util::EnsureAlignment(data, util::kValueAlignment, default_memory_pool()));

  ASSERT_OK(ValidateArrayFull(*aligned));
  AssertArraysEqual(*MakeArray(data), *MakeArray(aligned));
  this->CheckUnmodified(*data, *aligned);
}
```

Reply:
I added this test for all types.  It verifies that the array generated by `SampleArray` (i.e. an array generated with kDefaultBufferAlignment`) is suitably aligned for `kValueAlignment` and that `EnsureAlignment` does not modify it.